tracing-appender/src/non_blocking.rs

```diff
@@ -144,3 +144,124 @@ impl Drop for WorkerGuard {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rand::Rng;
+    use std::sync::Mutex;
+    use std::thread;
+    use std::time::Duration;
+
+    struct MockWriter {
+        writer: Arc<Mutex<Vec<String>>>,
+        max_writes_allowed: usize,
+        writes_attempted: usize,
+    }
+
+    impl Default for MockWriter {
+        fn default() -> Self {
+            MockWriter {
+                writer: Arc::new(Mutex::new(Vec::new())),
+                max_writes_allowed: 1,
+                writes_attempted: 0,
+            }
+        }
+    }
+
+    impl std::io::Write for MockWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let buf_len = buf.len();
+            self.writes_attempted += 1;
+            if self.writes_attempted > self.max_writes_allowed {
+                return Err(std::io::Error::from(std::io::ErrorKind::WouldBlock));
+            }
+
+            self.writer
+                .lock()
+                .expect("expected guard")
+                .push(String::from_utf8_lossy(buf).to_string());
+            Ok(buf_len)
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn logs_dropped_if_lossy() {
```

Comment:
We should have a test that when the writer is _not_ configured to be lossy, logs are not dropped. I think the alternative mock writer I proposed in https://github.com/tokio-rs/tracing/pull/678/files#r410371854 might help with this.

Reply:
Added a test